[mariadb] Pre-upgrade hook for db/user creation

fwiesel opened the pull request:
The idea is to roll-out new users with new credentials in a pre-upgrade hook
so that they are available to the actual deployment making use of it.

Users (or databases) cannot be deleted this way (much less rolled back).
The cleanup is left to a later iteration, which may make use of an expiry policy.

Instead of having custom init scripts with slightly different patterns,
the fields `databases` and `users` are there to create and update them.

This also allows for uniform access to other macros, such as side-car
containers, which need knowledge about the users and the databases as well.

By using a job which executes the change within the container
of the database, we can change the credentials of any user
without fearing a lockout.

The job will be executed on upgrade and rollback and since it
gets the values from the previous deployment, it not only
will set the new credentials on upgrade, it also restores the old ones
on rollback.

galkindmitrii commented on common/mariadb/templates/initdb/_init.sql.tpl:
```diff
@@ -1,3 +1,25 @@
+
+{{- if and .Values.global.dbUser .Values.global.dbPassword }}
+CREATE USER IF NOT EXISTS {{ .Values.global.dbUser }};
+GRANT ALL PRIVILEGES ON {{ .Values.name }}.* TO {{ .Values.global.dbUser }} IDENTIFIED BY {{ include "db_password" . }};
```
Do I understand that correct that in case of multiple Values.databases we are not adding the permissions?

fwiesel replied:
Yes, `.Values.databases` only creates the databases.
The users and their grants are created & maintained by `.Values.users` and `.Values.users[*].grants` respectively.

We can add some defaults, which avoids repetitive defaults, the issue though here is, that such a logic internal to this chart would be needed to replicated to anyone trying to evaluate the settings (i.e. the proxysql side-car,  PR #3554).

Certainly doable, but I think having to be explicit once in the parent chart is not a lot of effort. I would take care of it.

galkindmitrii commented on common/mariadb/values.yaml:
```diff
@@ -18,10 +18,29 @@ join_buffer_size: "4M"
+
+pre_change_hook:
+  image: "sapcc/unified-kubernetes-toolbox"
+  image_version: "20220525082919"
```
an extra image dependency, I assume the job will be failing on k8s upgrade and we would need extra coordination for deployment of all openstack services...

fwiesel replied:
I agree some extra work is needed on adapting the pipeline to make it work more seamlessly.
But I wouldn't say it requires extra coordination.  It uses the same image repository as in the ci-pipeline itself.

We need to update the pipelines (manually) when the image changes. That is exactly also the time we need to change this image version. So no extra coordination, but yes, extra manual effort.

We can change the pipeline-definition(s) to use exactly the same version, when the pipeline is applied.
Or we can change the pipeline to transport the latest version through the pipeline the same way all other images are transported.
Personally, I have a slight preference to the latter, but how it is done is a different discussion, I would say.

joker-at-work replied:
Just a remark: Not all users of the shared pipeline use image version transporting and not all users of this chart (I would assume) use the shared pipeline. Additionally, people tend to only update the shared tasks for a new image-version and not every reference, so Dmitry has a good point here.

fwiesel replied:
> Just a remark: Not all users of the shared pipeline use image version transporting and not all users of this chart (I would assume) use the shared pipeline

Well, but also the hook will only trigger if the "managed" database / user fields are used, so I don't think we have to cover all use-cases. Personally, I am fine with just the "openstack" services, or even nova and ironic, as they have more than one database & schema.

> Additionally, people tend to only update the shared tasks for a new image-version and not every reference, so Dmitry has a good point here.

I do not follow... 

I understand Dmitry the following way:
We upgrade k8s, so we need to ensure that the toolbox used everywhere is current enough.

There are 31 references to the `sapcc/unified-kubernetes-toolbox` in our pipeline definitions, 10 of those are under "shared", except in one location for ironic, all openstack services use the shared ones.

If we *need* to update the references in a pipeline then someone has to manually change those references,
and manually apply the pipeline. 

I am saying, the very procedure could result in the very same image being used the next time the pipeline is triggered inside this hooks as well in the pipeline job. 
Without effort of improving the situation, it just means that one has to change 32 places instead of 31.

Personally, I would opt for transporting the image version through the pipeline, and leave it to the teams which opt-out of image transport, but want to opt-in into the new functionality how they want to do it.

galkindmitrii commented on common/mariadb/templates/pre-change-job.yaml:
```diff
@@ -0,0 +1,41 @@
+      containers:
+      - name: pre-change-job
+        image: "{{ required ".Values.global.dockerHubMirrorAlternateRegion is missing" .Values.global.dockerHubMirrorAlternateRegion }}/{{ .Values.pre_change_hook.image }}:{{ .Values.pre_change_hook.image_version }}"
+        command:
```
Is there any chance we can do that easier?
e.g. by adding k8s lifecycle command with some failsafe sql? https://github.com/sapcc/helm-charts/blob/master/common/mariadb/templates/deployment.yaml#L55-L58

fwiesel replied:
I listed all the options in #3595. We probably move the discussions of alternatives there.

Just as a TLDR:
* `postStart` means restart. I want just want to get a user deployed out without a restart.
* While `postStart` running, the `Service` is down, as the container is [not marked running](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-state-running), so it adds to the downtime

fwiesel commented:
I've added PR  #3595 with the reasoning behind the decisions on how it has been realized.

joker-at-work commented:
I don't understand that sentence in commit "Pre-upgrade hook for db/user creation"

> Instead of having custom init scripts with sligthly different patterns,
> the fiels databases and users are there create and update the users.

same commit. this means "users" and not "uses", right? both could make sense, but create pretty different expectations

> which need knowledge about the uses

fwiesel commented:
> I don't understand that sentence in commit "Pre-upgrade hook for db/user creation"
> 
> > Instead of having custom init scripts with sligthly different patterns,
> > the fiels databases and users are there create and update the users.

Sorry, that was badly written, I've changed it to:
```
Instead of having custom init scripts with slightly different patterns,
the fields "databases" and "users" are there to create and update them.
```

> same commit. this means "users" and not "uses", right? both could make sense, but create pretty different expectations
> 
> > which need knowledge about the uses

Correct, I changed it.

IvoGoman commented on common/mariadb/templates/initdb/_init.sql.tpl:
```diff
@@ -1,3 +1,25 @@
-GRANT ALL PRIVILEGES ON {{.Values.name}}.* TO {{.Values.global.dbUser}}@localhost IDENTIFIED BY '{{include "db_password" .}}';
-GRANT ALL PRIVILEGES ON {{.Values.name}}.* TO {{.Values.global.dbUser}}@'%' IDENTIFIED BY '{{include "db_password" .}}';
+{{- if not .Values.databases }}
+CREATE DATABASE IF NOT EXISTS {{ .Values.name }};
```
The `CHARACTER SET utf8 COLLATE utf8_general_ci` should be left in.
Otherwise the default character set of the server will be used.
Checked this for Nova in QA and the server default is `latin1` but the nova db is created with `utf8`:
```
show variables like '%character_set_server%';
+----------------------+--------+
| Variable_name        | Value  |
+----------------------+--------+
| character_set_server | latin1 |
+----------------------+--------+
```
```suggestion
CREATE DATABASE IF NOT EXISTS {{ .Values.name }} CHARACTER SET utf8 COLLATE utf8_general_ci;
```

fwiesel replied:
In order to only have a single place where the character-set and collation is set, I added this in the front:
```
SET character_set_server = '{{ .Values.character_set_server }}';
SET collation_server = '{{ .Values.collation_server }}';
``` 

I hope, that's fine with you as well.

fwiesel replied:
No, I added it after your proposal.

IvoGoman commented on common/mariadb/templates/initdb/_init.sql.tpl:
```diff
@@ -1,3 +1,25 @@
+CREATE DATABASE IF NOT EXISTS {{ .Values.name }};
+{{- else }}
+    {{- range .Values.databases }}
+CREATE DATABASE IF NOT EXISTS {{ . }};
```
Same as above. `CHARACTER SET utf8 COLLATE utf8_general_ci` should be added to the create DB statement
```suggestion
CREATE DATABASE IF NOT EXISTS {{ . }} CHARACTER SET utf8 COLLATE utf8_general_ci;
```

fwiesel replied:
See [before](https://github.com/sapcc/helm-charts/pull/3555/files#r889121676)

fwiesel commented:
Hi, this PR is open now for three weeks. If there is no more feedback, I would assume that there is no strong objections, and I can merge it.

joker-at-work commented on common/mariadb/templates/pre-change-job.yaml:
```diff
@@ -0,0 +1,41 @@
+      - name: pre-change-job
+        image: "{{ required ".Values.global.dockerHubMirrorAlternateRegion is missing" .Values.global.dockerHubMirrorAlternateRegion }}/{{ .Values.pre_change_hook.image }}:{{ .Values.pre_change_hook.image_version }}"
+        command:
+        - kubectl-v{{ .Values.pre_change_hook.kubectl_version }}
```
Isn't `kubectl` in the image  wrapper that automatically uses the appropriate version? Or is this to guard against an old image being used and thus not having the version yet?

fwiesel replied:
The wrapper is for a different use-case ("out of cluster") and doesn't work with an in-cluster setup.

It requires a user and a password so it can do its magic (version discovery) and log-on to any kubernetes api endpoint and works with a kubeconfig.
The script relies on the in-cluster config and the credentials provided by the service-account.

Since we are now more strict on following upstream versions of kubernetes, I do not expect a large drift between versions across the pipeline, which requires an automatic version detection. So I think, we can avoid injecting the credentials and having an additional step (version discovery), which might fail. Still, it can be overwritten in each region, should the need arise.

joker-at-work commented on common/mariadb/templates/pre-change-job.yaml:
```diff
@@ -0,0 +1,41 @@
+        - exec
+        - -c
+        - mariadb
+        - deploy/{{ include "fullName" . }}
```
Why is this `deploy`? Is that short for `deployment`? If yes, why did you chose the short-form instead of the full name?

fwiesel replied:
> Why is this `deploy`? Is that short for `deployment`?

Yes:
```
$ kubectl api-resources | grep deploy
deployments                        deploy           apps/v1                                true         Deployment
```
> If yes, why did you chose the short-form instead of the full name?

That is the way it is done in the [official documentation](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#exec):
> kubectl exec deploy/mydeployment -- date

joker-at-work commented on common/mariadb/templates/pre-change-job.yaml:
```diff
@@ -0,0 +1,41 @@
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: "{{ include "fullName" . }}-pre-change"
```
Don't we need some special naming (a variable in the name e.g. chart version), because we're unable to update Jobs with helm3?

fwiesel replied:
No, the annotation `"helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded` takes care of that.
The job will be deleted when:
- the hook will be executed (so before the upgrade), in case it exists
- the hook has been successfully executed

In other words, helm should only leave the job around, if the hook didn't succeed, so one can debug it.
So one can simply retry the execution without having to manually delete the job, I've added `before-hook-creation` .

(Something which I think should also be added to https://github.com/sapcc/helm-charts/blob/master/common/region_check/templates/region-check-job.yaml#L16)

joker-at-work commented:
I'm fine with merging this, but would like to see @viktor-kovacs and @bashar-alkhateeb opinion.

fwiesel commented:
@viktor-kovacs @bashar-alkhateeb What is your view on this PR?

viktor-kovacs commented:
> @viktor-kovacs @bashar-alkhateeb What is your view on this PR?

I'm fine with this PR too.

bashar-alkhateeb commented:
somehow it sounds odd, to spin up a container (in a job) and give it a service account to configure another container in runtime using kubectl command.
i would prefer to have a precess inside the mariadb container to reconfigure it in case the config map got updated. (watching the configmap update)

fwiesel commented:
> i would prefer to have a process inside the mariadb container to reconfigure it in case the config map got updated. (watching the configmap update)

Well, I tried that with rabbitmq and that had some negative feedback.
See the [ADR](https://github.com/sapcc/helm-charts/pull/3595) for the reasoning.
I want to avoid to have the same discussion in multiple places so I would like to focus that there.

For me more important which option to chose it to have an implementation in place.
We can still improve on that.

fwiesel commented:
@bashar-alkhateeb Is that more an remark, or are you opposing this change?
